core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java

```diff
@@ -466,700 +619,784 @@ private void buildChannel() {
+         for (View group : jgroupsSubGroups) {
+            subGroups.add(fromJGroupsAddressList(group.getMembers()));
+         }
+      } else {
+         CLUSTER.receivedClusterView(channel.clusterName(), new_view);
+         subGroups = Collections.emptyList();
+      }
+      receiveClusterView(new_view.getViewId().getId(), fromJGroupsAddressList(new_view.getMembers()), subGroups);
    }
 
-   @Override
-   public Address getCoordinator() {
-      return coordinator;
+   private static List<Address> fromJGroupsAddressList(List<org.jgroups.Address> list) {
+      return Collections.unmodifiableList(list.stream()
+                                              .map(JGroupsAddressCache::fromJGroupsAddress)
+                                              .collect(Collectors.toList()));
    }
 
-   private void waitForInitialNodes() {
-      int initialClusterSize = configuration.transport().initialClusterSize();
-      if (initialClusterSize <= 1)
+   public void receiveClusterView(long viewId, List<Address> members, List<List<Address>> subgroups) {
+      if (members.isEmpty()) {
          return;
+      }
 
-      long timeout = configuration.transport().initialClusterTimeout();
-      long remainingNanos = TimeUnit.MILLISECONDS.toNanos(timeout);
+      ClusterView oldView = this.clusterView;
+
+      // Update every view-related field while holding the lock so that waitForView only returns
+      // after everything was updated.
+      CompletableFuture<Void> oldFuture = null;
       viewUpdateLock.lock();
       try {
-         while (channel != null && channel.getView().getMembers().size() < initialClusterSize &&
-               remainingNanos > 0) {
-            log.debugf("Waiting for %d nodes, current view has %d", initialClusterSize,
-                  channel.getView().getMembers().size());
-            remainingNanos = viewUpdateCondition.awaitNanos(remainingNanos);
+         // Delta view debug log for large cluster
+         if (log.isDebugEnabled() && oldView.getMembers() != null) {
+            List<Address> joined = new ArrayList<>(members);
+            joined.removeAll(oldView.getMembers());
+            List<Address> left = new ArrayList<>(oldView.getMembers());
+            left.removeAll(members);
+            log.debugf("Joined: %s, Left: %s", joined, left);
          }
-      } catch (InterruptedException e) {
-         log.interruptedWaitingForCoordinator(e);
-         Thread.currentThread().interrupt();
-      } finally {
-         viewUpdateLock.unlock();
-      }
 
-      if (remainingNanos <= 0) {
-         throw log.timeoutWaitingForInitialNodes(initialClusterSize, channel.getView().getMembers());
-      }
-
-      log.debugf("Initial cluster size of %d nodes reached", initialClusterSize);
-   }
+         // The first view is installed before returning from JChannel.connect
+         // So we need to set the local address here
+         if (address == null) {
```

Comment:
the other `receiveClusterView()` already sets the address before invoking this method.